### PR TITLE
gitstatus modular

### LINKED ISSUE #1831: Interested in refactoring $PROMPT format to work like a formatted string
I am interested in working on  a refactor of how the prompt string is generated, but I wanted to ask if it was of interest or not before beginning to work on it.

In particular, I feel like most users of xonsh will be familiar with typical string `format` functionality, but that is not how the `$PROMPT` is actually generated. It appears to be pieced together in stages and therefore it is difficult to grasp at first how the `$FORMATTER_DICT` will actually behave until after playing with it for awhile. Instead, I propose that the `$FORMATTER_DICT` be renamed to renamed to `$PROMPT_FORMATTER` and be used to generate a dictionary that is passed to `$PROMPT.format(**generated_kwargs)` to make it as transparent as possible. 

An example of how I see the `$PROMPT` variable changing:

```
# current default
$PROMPT = '{env_name:{} }{BOLD_GREEN}{user}@{hostname}{BOLD_BLUE} {cwd}{branch_color}{curr_branch: {}}{NO_COLOR} {BOLD_BLUE}{prompt_end}{NO_COLOR}

# proposed change
$PROMPT = '{env_name:{{%e}} }{color:bold_green}{user}@{hostname}{color:bold_blue} {cwd}{branch_color}{curr_branch:{{%b}}}{color:no_color} {color:bold_blue}{prompt_end}{color:no_color}
```

The proposed items in `$PROMPT_FORMATTER` will include both callables and objects that implement the `__format__` method to produce customizable output. A prime example use case would be a color object that produces the appropriate color escape sequence when passed the color string to `__format__`. This would also give us the ability to highly customize the way that `$PROMPT_FORMATTER['gitstatus']` looks. If a callable fails or throws an exception, we could replace it with a `PromptError` object that simply prints that the exception occurred.

I wanted to propose this and see if there is interest before going to far. Also, are there ways to set `pdb` traces inside xonsh running framework. I was unable to get the trace to fire in my initial testing.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by jnoortheen**
This is created after needing to reuse and extend gitstatus from prompt function in efficient way. 
Possibly we can close #1831 #1756 #4493 after this

## Todo:

- [x] add docs

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

File: xonsh/prompt/base.py [hunk 1]
```diff
@@ -293,3 +258,214 @@ def _format_value(val, spec, conv):
...
+
+        value = self[name]
+        return isinstance(value, BasePromptFld) or callable(value)
+
+    def reset(self):
+        """the results are cached and need to be reset between prompts"""
+        self._cache.clear()
+
+
+class BasePromptFld:
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/prompt/base.py:379]: I am not opposed to that. I prefer shorter names when they are expressive enough.

**Review by gforsyth — CHANGES_REQUESTED** [review]
Hey @jnoortheen -- thanks for putting this together, it seems like a big improvement!
I've called out a few nits in the code -- I also have two asks for this:
1. Can we make the `repr` of `PROMPT_FIELDS` a little friendlier?  Maybe `list($PROMPT_FIELDS.items())`  or similar?
2. Some docs on how to customize the prompt, especially bits of `gitstatus`

File: xonsh/prompt/base.py [hunk 2]
```diff
@@ -293,3 +258,214 @@ def _format_value(val, spec, conv):
...
+            self.load_initial()
+
+    def __getitem__(self, item: "str|BasePromptFld"):
+        # todo: load on-demand from modules
+        # self._matcher = ModuleFinder(
+        #     "xonsh.prompts",
+        #     *XSH.env.get("XONSH_PROMPT_NS", []),
+        # )
+        # if item not in self._items:
+        #     self._matcher.get_module(item)
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/prompt/base.py]: done

File: xonsh/prompt/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/prompt/base.py]: done

**Comment by jnoortheen** [comment]
@gforsyth I have added docs to the gitstatus module

**Pull request merged**